### PR TITLE
Pass configured/auth headers to a Datasource.

### PR DESCRIPTION
**Pull request by mrsiano**
In some setups (ex openshift), the Datasource will require Grafana
to pass any headers when sending queries. Which is something
Grafana currently does not support.

https://github.com/grafana/grafana/issues/10971

### CODE REVIEW

**Review by torkelo — CHANGES_REQUESTED** [review]
Added the mrsiano to the CLA signed list, the bot is rechecking PRs now, should pass that test. 

This PR needs some more work

1) Rebase on master, fix compilation errors
2) Need a more generic solution, not a specific one for Authorization header only.

**Comment by mrsiano** [comment]
@torkelo @bergquist I've update the PR with more generic solution.
currently I tried to allow pass any custom header in the json data.
we should separate multiple headers with ",".
we can change it by having more textbox in the ui as the end user wish, but it's something we can have by another PR later on.

File: pkg/api/pluginproxy/ds_proxy.go
Comment (torkelo) [inline pkg/api/pluginproxy/ds_proxy.go]: misspelling of costum :)

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (torkelo) [inline public/app/features/plugins/partials/ds_http_settings.html]: this option is no longer needed

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (torkelo) [inline public/app/features/plugins/partials/ds_http_settings.html]: think this should be an array with an action button "Add custom header" (so current.jsonData.customHeaders is an array)

**Review by torkelo — DISMISSED** [review]
Making progress, still needs some minor changes, and a rebase on top of master (So tests pass)

File: pkg/api/pluginproxy/ds_proxy.go
Comment (bergquist) [inline pkg/api/pluginproxy/ds_proxy.go]: This should be debug level and don't log the header value. Token's are as secret as passwords.

File: pkg/api/pluginproxy/ds_proxy.go
Comment (mrsiano) [reply pkg/api/pluginproxy/ds_proxy.go]: I wanted to be custom like "tailor made" but not "cast"
let's keep it ?!

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (mrsiano) [reply public/app/features/plugins/partials/ds_http_settings.html]: done

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (mrsiano) [reply public/app/features/plugins/partials/ds_http_settings.html]: ok.

File: pkg/api/pluginproxy/ds_proxy.go
Comment (mrsiano) [reply pkg/api/pluginproxy/ds_proxy.go]: done

File: pkg/api/pluginproxy/ds_proxy.go
Comment (bergquist) [reply pkg/api/pluginproxy/ds_proxy.go]: Should be `custom`

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (mrsiano) [reply public/app/features/plugins/partials/ds_http_settings.html]: This is a tricky area since secureJsonData expect string as value.
we can change it to interface{} but it means more code change.

using JsonData could be another option, but required more tweaks and code changes in diffrent  areas.

any thoughts ?

@torkelo @bergquist

Comment (torkelo) [reply public/app/features/plugins/partials/ds_http_settings.html]: No that should not be required as secureJsonData is a map (key value with string values) so just put header key names in jsonData.header1 jsonData.header2 and values in secureJsonData.header1 etc

Comment (bergquist) [reply public/app/features/plugins/partials/ds_http_settings.html]: Using `jsonData` for header names and `securejsonData` for values sounds like a good idea. I suggest you prefix the header key and value with something to make it easier to loop over them.

Comment (mrsiano) [reply public/app/features/plugins/partials/ds_http_settings.html]: yep, i'm about to use ng-repeater to iterate them.

Comment (mrsiano) [reply public/app/features/plugins/partials/ds_http_settings.html]: can you suggest how to retrive the $scope ds_edit_ctrl.

I wanted to implement somthing similar to this concept
http://jsfiddle.net/yhv9bjrx/2/

Comment (bergquist) [reply public/app/features/plugins/partials/ds_http_settings.html]: @mrsiano Try adding a functions in https://github.com/grafana/grafana/blob/master/public/app/features/plugins/ds_edit_ctrl.ts for returning the data.

**Comment by mrsiano** [comment]
PR has been updated, PTAL
@bergquist

**Comment by mrsiano** [comment]
@bergquist @torkelo 
when trying to use that PR with curl sending the secureJsonData 
the backend didn't see the secureJsonData, we need to convert it somehow?!

when create a DS from the UI it works fine.
`
 {"datasource":{"id":45,"orgId":1,"name":"prodev","type":"prometheus","typeLogoUrl":"","access":"proxy","url":"server,"password":"","user":"","database":"","basicAuth":false,"basicAuthUser":"","basicAuthPassword":"","withCredentials":false,"isDefault":false,"jsonData":{"tlsSkipVerify":true},"secureJsonFields":{"customHeaders":true},"version":1,"readOnly":false},"id":45,"message":"Datasource added","name":"prodev"}
`

**Comment by brancz** [comment]
Any chance that we can have this be read from a file and configured through provisioning?

**Comment by mrsiano** [comment]
yep, actually this is one of the reasons we submitted this PR to be able to run DS through provisioning.
if so can we merge it ?!

**Comment by brancz** [comment]
It would be nice if all we needed was a `bearer_token_file` entry that would read it from the file and use the token it has. Otherwise it's non trivial to run Grafana on Kubernetes, and we again need some sidecar that templates the serviceaccount token into the datasource file used for provisioning.

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (bergquist) [inline public/app/features/plugins/partials/ds_http_settings.html]: the width for the gf-form-switch should be 11. and if should be last in the list of form-switches.

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (bergquist) [inline public/app/features/plugins/partials/ds_http_settings.html]: Its impossible to add new headers since the add button is placed within the repeater. If the list is empty nothing shows up.

File: public/app/features/plugins/partials/ds_http_settings.html
Comment (bergquist) [inline public/app/features/plugins/partials/ds_http_settings.html]: this controller does not exist. So it raises an error. I don't think we should create a new controller for this. Just reuse the `DataSourceEditCtrl`

File: pkg/api/pluginproxy/ds_proxy.go
Comment (bergquist) [inline pkg/api/pluginproxy/ds_proxy.go]: still logging the header value

File: pkg/api/pluginproxy/ds_proxy.go
Comment (bergquist) [inline pkg/api/pluginproxy/ds_proxy.go]: Should be named `CustomHeader`

File: pkg/api/pluginproxy/ds_proxy.go
Comment (bergquist) [inline pkg/api/pluginproxy/ds_proxy.go]: Please run `gofmt -w -s ./pkg` in the grafana root to fix the build errors.

**Review by bergquist — CHANGES_REQUESTED** [review]
Still some changes require before we merge this. I could fix them but Proerbly not a week or so. So I suggest you fix them if you want this to be merge for that.

**Comment by bergquist** [comment]
To have good support for tokens provided by k8s we might have to rethink this a little bit. 

```
09:07 <brancz> basically in order to get the bearer token into the provisioning file we would need additional code
09:08 <brancz> concretely the service account token is located in /var/run/kubernetes.io/serviceaccount/token
09:09 <brancz> in Prometheus what we did is instead of requiring the plaintext token to be in the config file, we allow specifying just the file to be read for the token
```

**Comment by mrsiano** [comment]
@bergquist right, but we want this PR to support generic headers.
and not k8s in particular.

seems like we can have another PR to read the k8s token file.
also we probably gonna need to support k8s and openshift.

@jeremyeder @brancz

**Comment by brancz** [comment]
Note that what I suggested isn't a hard Kubernetes integration, it merely happens to work with Kubernetes/OpenShift and other configuration/secret management systems (like hashicorp vault). Yes, I think a follow up PR would be nice, where a `bearer_token_file` field in the provisioning files could be a mean to populate the "custom headers" proposed in this PR.

File: pkg/api/pluginproxy/ds_proxy.go
Comment (mrsiano) [reply pkg/api/pluginproxy/ds_proxy.go]: thats require structure change the SecureJsonData is map[string]string..
any thoughts?

Comment (torkelo) [reply pkg/api/pluginproxy/ds_proxy.go]: No the code will be cleaner :) no need to parse or have multiple values, just single value for each key. Super simple and clean.

Comment (mrsiano) [reply pkg/api/pluginproxy/ds_proxy.go]: Can you suggest how to do that in the ds_http_settings.html
I tried to index the header[$index]

but than it adds as dict, so we'll need to decode it, in order to retrive the index 

also, Is it possible to have a follow up PR for that ?

Comment (torkelo) [reply pkg/api/pluginproxy/ds_proxy.go]: You Will have to write view model handled the translation , don’t find the storage model directly in the ui

Comment (torkelo) [reply pkg/api/pluginproxy/ds_proxy.go]: You Will have to write view model handled the translation , don’t find the storage model directly in the ui

Comment (mrsiano) [reply pkg/api/pluginproxy/ds_proxy.go]: is that necessary, or we can have a follow up PR ? 
also, we can make that PR work only for provisioning

Comment (mrsiano) [reply pkg/api/pluginproxy/ds_proxy.go]: whats the point to store them speratly that will force us to unmarshal the jsonData in the back
the following peace of code it represting dealing only with secureJson
`
 		// Parse Custom Header
		if proxy.ds.SecureJsonData != nil {
			decryptSdj := proxy.ds.SecureJsonData.Decrypt()
			logger.Info("debug", "", decryptSdj)

			index := 1
			for {
				headerNameSuffix := fmt.Sprintf("httpHeaderName%d", index)
				headerValueSuffix := fmt.Sprintf("httpHeaderValue%d", index)

				if key, ok := decryptSdj[headerNameSuffix]; ok {
					if val, ok := decryptSdj[headerValueSuffix]; ok {
						req.Header.Add(key, val)
						logger.Debug("Using custom header ", "CustomHeaders", key)
					}
				} else {
					break
				}
				index += 1
			}
		}
`

and here is who I use it in the datasource yaml file
`
   secureJsonData:
      httpHeaderName1: "Authorization"
      httpHeaderValue1: " ... "
`

Comment (torkelo) [reply pkg/api/pluginproxy/ds_proxy.go]: The point is to be able to show in the UI what headers are set (as the secureJson fields are not available decrypted to the frontend)

**Comment by mrsiano** [comment]
@torkelo @bergquist any rejections ?! otherwise, we can approve it?

**Comment by torkelo** [comment]
Looks a lot better now! Thanks for keeping up with all our change requests. The only minor issue is that you cannot see what headers you added in the UI (not even possible to implement) as the header name is also encrypted and the secureJson fields are never exposed to be browser. So think it makes sense to have the header names in the unencrypted part.

**Comment by mrsiano** [comment]
@torkelo PTAL

File: pkg/api/pluginproxy/ds_proxy.go
Comment (bergquist) [inline pkg/api/pluginproxy/ds_proxy.go]: This will cause an infinite loop since `.Get` will never return `nil`
https://github.com/grafana/grafana/blob/master/pkg/components/simplejson/simplejson.go#L148

Please extract this piece of code into a new function and add tests for this feature.

**Review by bergquist — CHANGES_REQUESTED** [review]
This code creates an inifinite loop which needs to be fixed.

Please rebase this PR onto master to solve the codepsell issues. 

Adding this only using provisioning sounds like a good idea.

**Comment by mrsiano** [comment]
@bergquist PTAL

File: pkg/api/pluginproxy/ds_proxy_test.go
Comment (mrsiano) [inline pkg/api/pluginproxy/ds_proxy_test.go]: @bergquist PTAL

**Review by bergquist — APPROVED** [review]

**Comment by bergquist** [comment]
Great! This looks good.

@mrsiano it still seems like our CLA bot cannot identify you. Could you please add your gmail as a secondary email to your github account? Otherwise, the CLA bot cannot connect you and the commit.

File: pkg/api/pluginproxy/ds_proxy_test.go
Comment (bergquist) [reply pkg/api/pluginproxy/ds_proxy_test.go]: LGTM

**Comment by mrsiano** [comment]
@bergquist I've already did that, but from some reason it didn;t pass it, @torkelo already add me

**Comment by bergquist** [comment]
could you try making it public for while? I cant see the email on your profile so I guess the bot be able to find it either.

**Comment by bergquist** [comment]
woho! Now the CLA bot is happy :) Will merge once all build passes.

**Comment by bergquist** [comment]
Follow up issue for reading the config value from file https://github.com/grafana/grafana/issues/12368

**Comment by jeremyeder** [comment]
@bergquist thank you for your patient reviews of this.

**Comment by tonglil** [comment]
Would there be a possibility to accept a PR that allows configuring the header value from a file (instead of hard code in provisioning config)?

In our use case, the header value is an oauth token that will need to be refreshed. We are able to write that token to a file, which we would like to use as a header's value.

**Comment by brancz** [comment]
@tonglil the follow up issue with what you describe has already been opened: https://github.com/grafana/grafana/issues/12368

**Comment by braggjc** [comment]
Hey guys I have pulled down the grafana:5.3.0-beta1 docker image and have manually loaded my new elasticsearch datasource with the customerHeader tags however, I am still getting a **502** error when attempting to "Save & Test" it in the application.   This would be connecting to the Openshift Origin 3.9 environment, that has a standard EFK stack for log aggregation. 

```
{
	"orgId": 1,
	"name": "Elasticsearch2",
	"type": "elasticsearch",
	"typeLogoUrl": "public/app/plugins/datasource/elasticsearch/img/elasticsearch.svg",
	"access": "proxy",
	"url": "http://es.ocp-dc-na1-nonprod.myworld.com:443",
	"password": "",
	"user": "",
	"database": "[project.abc-api-test.*].YYYY.MM.DD",
	"basicAuth": false,
	"isDefault": false,
	"jsonData": {
		"esVersion": 5,
		"interval": "Daily",
		"keepCookies": [],
		"maxConcurrentShardRequests": 256,
		"timeField": "@timestamp",
		"tenant": "abc-api-test",
		"httpHeaderName1": "Authorization",
		"httpHeaderName2": "X-Proxy-Remote-User",
		"tlsAuth": false,
		"tlsAuthWithCACert": false,
		"tlsSkipVerify": true
	},
	"secureJsonData": {
		"httpHeaderValue1": "Bearer uDAfrDnvi08QfuP22BWQNxso5LH3Q1iWYFvcRb_yQ70",
		"httpHeaderValue2": "jbragg"
	},
	"readOnly": false
}
```

Checking logs I see:
```
...
2018/09/20 13:43:18 http: proxy error: EOF
t=2018-09-20T13:43:18+0000 lvl=info msg="Request Completed" logger=context userId=1 orgId=1 uname=admin method=GET path=/api/datasources/proxy/41/project.abc-api-test.*.2018.09.20/_mapping status=502 remote_addr=172.17.0.1 time_ms=159 size=0 referer=http://localhost:3000/datasources/edit/41
...
```

I am able to successfully invoked directly via curl:
```
curl -o curlout.json -k -H "X-Proxy-Remote-User: jbragg" -H "Authorization: Bearer uDAfrDnvi08QfuP22BWQNxso5LH3Q1iWYFvcRb_yQ70" "https://es.ocp-dc-na1-nonprod.myworld.com/project.abc-api-test.*.2018.09.20/_mapping"
```

Is there something else I am missing?

**Comment by braggjc** [comment]
> Hey guys I have pulled down the grafana:5.3.0-beta1 docker image and have manually loaded my new elasticsearch datasource with the customerHeader tags however, I am still getting a **502** error when attempting to "Save & Test" it in the application. This would be connecting to the Openshift Origin 3.9 environment, that has a standard EFK stack for log aggregation.
> 
> ```
> {
> 	"orgId": 1,
> 	"name": "Elasticsearch2",
> 	"type": "elasticsearch",
> 	"typeLogoUrl": "public/app/plugins/datasource/elasticsearch/img/elasticsearch.svg",
> 	"access": "proxy",
> 	"url": "http://es.ocp-dc-na1-nonprod.myworld.com:443",
> 	"password": "",
> 	"user": "",
> 	"database": "[project.abc-api-test.*].YYYY.MM.DD",
> 	"basicAuth": false,
> 	"isDefault": false,
> 	"jsonData": {
> 		"esVersion": 5,
> 		"interval": "Daily",
> 		"keepCookies": [],
> 		"maxConcurrentShardRequests": 256,
> 		"timeField": "@timestamp",
> 		"tenant": "abc-api-test",
> 		"httpHeaderName1": "Authorization",
> 		"httpHeaderName2": "X-Proxy-Remote-User",
> 		"tlsAuth": false,
> 		"tlsAuthWithCACert": false,
> 		"tlsSkipVerify": true
> 	},
> 	"secureJsonData": {
> 		"httpHeaderValue1": "Bearer uDAfrDnvi08QfuP22BWQNxso5LH3Q1iWYFvcRb_yQ70",
> 		"httpHeaderValue2": "jbragg"
> 	},
> 	"readOnly": false
> }
> ```
> Checking logs I see:
> 
> ```
> ...
> 2018/09/20 13:43:18 http: proxy error: EOF
> t=2018-09-20T13:43:18+0000 lvl=info msg="Request Completed" logger=context userId=1 orgId=1 uname=admin method=GET path=/api/datasources/proxy/41/project.abc-api-test.*.2018.09.20/_mapping status=502 remote_addr=172.17.0.1 time_ms=159 size=0 referer=http://localhost:3000/datasources/edit/41
> ...
> ```
> I am able to successfully invoked directly via curl:
> 
> ```
> curl -o curlout.json -k -H "X-Proxy-Remote-User: jbragg" -H "Authorization: Bearer uDAfrDnvi08QfuP22BWQNxso5LH3Q1iWYFvcRb_yQ70" "https://es.ocp-dc-na1-nonprod.myworld.com/project.abc-api-test.*.2018.09.20/_mapping"
> ```
> Is there something else I am missing?

Please ignore: I noticed after posting that I was not technically using the correct protocol for my url, should simple be: ("url": "**https**://es.ocp-dc-na1-nonprod.myworld.com",)

**Pull request merged**
